### PR TITLE
feat: cfd-481 - fix update fare triangle

### PR DESCRIPTION
**Pull request by JwanKhalaf**
# Description

-   Implement the missed acceptance criteria and refactored code.

# Testing instructions

-   From `/products/productDetails` page, attempt to edit fare triangle and edit the CSV to ensure error handling is working (e.g. edit csv by adding new fare stage, and also test by changing the name on one of the fare stages but keep the same number of stages)

# Type of change

-   [x] feat - A new feature
-   [ ] fix - A bug fix
-   [ ] docs - Documentation only changes
-   [ ] codestyle - Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
-   [ ] refactor - A code change that neither fixes a bug nor adds a feature
-   [ ] perf - A code change that improves performance
-   [ ] test - Adding missing tests or correcting existing tests
-   [ ] build - Changes that affect the build system or external dependencies (example scopes: gulp, broccoli, npm)
-   [ ] ci - Changes to our CI configuration files and scripts (example scopes: Travis, Circle, BrowserStack, SauceLabs)
-   [ ] chore - Other changes that don't modify src or test files
-   [ ] revert - Reverts a previous commit
-   [ ] content - Changes to content or copy

### CODE REVIEW

File: repos/fdbt-site/src/constants/attributes.ts [hunk 1]
```diff
@@ -112,4 +112,4 @@ export const CSV_ZONE_FILE_NAME = 'fdbt-csv-zone-file-name';
 
 export const MATCHING_JSON_ATTRIBUTE = 'fdbt-matching-json';
 
-export const PRODUCT_AND_SERVICE_ID_ATTRIBUTE = 'fdbt-product-and-service-id';
+export const MATCHING_JSON_META_DATA_ATTRIBUTE = 'fdbt-product-and-service-id';
```
Comment (RobCatton) [inline repos/fdbt-site/src/constants/attributes.ts:115]: Should we change the string value of this to something more appropriate for the attribute?

File: repos/fdbt-site/src/pages/api/csvUpload.ts [hunk 2]
```diff
@@ -198,15 +209,54 @@ export const faresTriangleDataMapper = (
...
+        const combinedNamesOnly = combinedFareZones.map((x) => x.name);
+
+        // get rid of duplicates
+        fareZoneNames = [...new Set(combinedNamesOnly)];
+    }
+
+    return fareZoneNames;
+};
+
+export const thereIsAFareStageNameMismatch = (fareTriangleData: UserFareStages, fareZoneNames: string[]): boolean => {
```
Comment (RobCatton) [inline repos/fdbt-site/src/pages/api/csvUpload.ts:229]: Could we do it like this? 
```suggestion
export const thereIsAFareStageNameMismatch = (fareTriangleData: UserFareStages, fareZoneNames: string[]): boolean => fareTriangleData.fareStages.some((fs) => {
        return !fareZoneNames.includes(fs.stageName);
    });

```

File: repos/fdbt-site/src/pages/api/csvUpload.ts [hunk 3]
```diff
@@ -218,32 +268,116 @@ export default async (req: NextApiRequestWithSession, res: NextApiResponse): Pro
...
+                }
+
+                // put the now updated matching json into s3
+                // overriding the existing object
+                await putUserDataInProductsBucketWithFilePath(ticket, matchingJsonMetaData.matchingJsonLink);
+
+                redirectTo(
+                    res,
+                    `/products/productDetails?productId=${matchingJsonMetaData?.productId}&serviceId=${matchingJsonMetaData?.serviceId}`,
+                );
```
Comment (RobCatton) [inline repos/fdbt-site/src/pages/api/csvUpload.ts:352]: I think we need a return after here, as it's a redirect.

File: repos/fdbt-site/src/pages/api/csvUpload.ts (hunk 2 above)
Comment (RobCatton) [reply repos/fdbt-site/src/pages/api/csvUpload.ts:229]:
```suggestion
export const thereIsAFareStageNameMismatch = (fareTriangleData: UserFareStages, fareZoneNames: string[]): boolean =>
    fareTriangleData.fareStages.some((fs) => {
        return !fareZoneNames.includes(fs.stageName);
    });

```

File: repos/fdbt-site/src/pages/api/csvUpload.ts (hunk 3 above)
Comment (JwanKhalaf) [reply repos/fdbt-site/src/pages/api/csvUpload.ts:352]: Do we? it didn't have a `return` after the redirect before?

**Review by RobCatton — APPROVED** [review]

**Pull request merged**